src/util/serializer.h

```diff
@@ -0,0 +1,818 @@
+#pragma once
+
+#ifndef CPROVER_UTIL_SERIALIZER_H
+#define CPROVER_UTIL_SERIALIZER_H
+
+#include <utility>
+#include <string>
+#include <functional>
+#include <vector>
+#include <set>
+#include <map>
+#ifdef USE_BOOST
+#include <boost/bimap.hpp>
+#endif
+
+
+
+/*******************************************************************\
+
+Class: serializer_traitst
+
+Purpose:
+  Defines traits attached to a particular serializer.
+  This is used to augment the behaviour of a family of serializers, typically
+  by passing data that is required for serializing a particular type of object
+  through to a sub-object serialization routine.
+
+\*******************************************************************/
+class serializer_traitst
+{
+public:
+  // The virtual destructor ensures sub-classes are disposed correctly.
+  virtual ~serializer_traitst() = default;
+};
+
+class serializert
+{
+private:
+  // The serializer for a containing object, if any
+  serializert *parent;
+  // Whether this serializer is used for reading rather than writing
+  bool is_read;
+  // Traits attached to this serializer
+  serializer_traitst *traits;
+
+protected:
+  /*******************************************************************\
+
+  Function: serializert::serializert
+
+  Inputs:
+    parent: The serializer for the containing object.
+    is_read: Whether this serializer is used for reading rather than writing.
+
+  Outputs:
+
+  Purpose:
+    Creates a serializert for a sub-object of a larger serialization.
+    This is only called internally by sub-class implementations.
+
+  \*******************************************************************/
+  serializert(serializert *parent, bool is_read)
+    : parent(parent), is_read(is_read)
+  {
+  }
+
+  /*******************************************************************\
+
+  Function: serializert::serializert
+
+  Inputs:
+    is_read: Whether this serializer is used for reading rather than writing.
+
+  Outputs:
+
+  Purpose:
+    Creates a serializert.
+    This will be exposed externally by sub-class implementations but is not
+    itself public.
+
+  \*******************************************************************/
+  explicit serializert(bool is_read)
+    : parent(nullptr), is_read(is_read)
+  {
+  }
+
+  // The virtual destructor ensures sub-classes are disposed correctly.
+  virtual ~serializert() = default;
+
+public:
```

Comment:
merge public sections (resp. protected, private) so that it is easier to see which methods are in the interface of the class